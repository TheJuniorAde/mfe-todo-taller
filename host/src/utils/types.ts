export interface TodoContextState {
  list: TodoItem[]
  updateData: (newTodo: TodoItem) => void
  toggleStatus: (newTodo: TodoItem) => void
}

export interface TodoItem {
  id: string
  description: string
  status: "completed" | "incomplete"
  finishedAt: string
  createdAt: string
}
