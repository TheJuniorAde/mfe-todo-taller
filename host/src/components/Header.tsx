import React from "react"
import { NavLink } from "react-router-dom"

export const Header: React.FC = () => (
  <div className="p-5 bg-blue-500 text-white -text-3xl font-bold">
    <NavLink
      data-testid="to-list"
      className="px-3 inline-block align-baseline font-bold text-sm"
      to="/"
    >
      List All Todo
    </NavLink>
    <NavLink
      data-testid="to-form"
      className="px-3 inline-block align-baseline font-bold text-sm"
      to="/create"
    >
      Create Todo Item
    </NavLink>
  </div>
)
