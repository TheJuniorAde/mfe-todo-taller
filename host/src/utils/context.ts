import { createContext } from "react"
import { v4 } from "uuid"
import { TodoContextState, TodoItem } from "./types"

export const TodoContext = createContext<TodoContextState>({
  list: [],
  updateData: () => null,
  toggleStatus: () => null,
})

export const loadStorageData = (): TodoItem[] => {
  const localData = localStorage.getItem("todo-list")
  let updatedData: TodoItem[] = []

  if (!localData || !String(localData).trim()) return []

  try {
    updatedData = JSON.parse(localData)
  } catch (_) {
    console.log("unable to parse the data")
  }

  return updatedData
}

export const updateData = (newTodo: TodoItem) => {
  const newList = [...loadStorageData()]
  const todoIndex = newList.findIndex((curTodo) => curTodo.id === newTodo.id)

  if (todoIndex === -1) {
    newList.push({
      ...newTodo,
      createdAt: new Date().toLocaleString(),
      id: v4(),
    })
  } else {
    newList[todoIndex] = { ...newList[todoIndex], ...newTodo }
  }

  localStorage.setItem("todo-list", JSON.stringify(newList))
}

export const toggleStatus = (todo: TodoItem) =>
  updateData({
    ...todo,
    status: todo.status === "incomplete" ? "completed" : "incomplete",
    finishedAt: todo.status === "incomplete" ? new Date().toLocaleString() : "",
  })
