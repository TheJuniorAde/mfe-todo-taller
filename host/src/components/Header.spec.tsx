/**
 * @jest-environment jsdom
 */
import React from "react"
import { render } from "@testing-library/react"
import { Header } from "./Header"
import { TodoContext } from "../utils"
import { BrowserRouter } from "react-router-dom"

describe("Header", () => {
  it("renders correctly", () => {
    const { container } = render(
      <BrowserRouter>
        <TodoContext.Provider
          value={{ list: [], updateData: jest.fn(), toggleStatus: jest.fn() }}
        >
          <Header />
        </TodoContext.Provider>
      </BrowserRouter>
    )

    expect(container.parentElement).toMatchSnapshot()
  })
})
