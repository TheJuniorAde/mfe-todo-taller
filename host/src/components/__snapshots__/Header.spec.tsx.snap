// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<body>
  <div>
    <div
      class="p-5 bg-blue-500 text-white -text-3xl font-bold"
    >
      <a
        aria-current="page"
        class="px-3 inline-block align-baseline font-bold text-sm active"
        data-testid="to-list"
        href="/"
      >
        List All Todo
      </a>
      <a
        class="px-3 inline-block align-baseline font-bold text-sm"
        data-testid="to-form"
        href="/create"
      >
        Create Todo Item
      </a>
    </div>
  </div>
</body>
`;
